File: server/internal/control/agent_engines.go
package control

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"
)

const (
	agentEngineCodex      = "codex"
	agentEngineClaudeCode = "claude_code"
	agentEnginePi         = "pi"
)

type AgentEngineCatalogItem struct {
	ID         string `json:"id"`
	Name       string `json:"name"`
	Mention    string `json:"mention"`
	Capability string `json:"capability"`
}

func fixedAgentEngineCatalog() []AgentEngineCatalogItem {
	return []AgentEngineCatalogItem{
		{ID: agentEngineCodex, Name: "Codex", Mention: "@codex", Capability: "codex"},
		{ID: agentEngineClaudeCode, Name: "Claude Code", Mention: "@claude", Capability: "claudeCode"},
		{ID: agentEnginePi, Name: "Pi", Mention: "@pi", Capability: "pi"},
	}
}

func normalizeAgentEngineInput(agentEngine, legacyProvider, legacyAgentProfile string) (string, error) {
	if engine := strings.ToLower(strings.TrimSpace(agentEngine)); engine != "" {
		if !isSupportedAgentEngine(engine) {
			return "", errors.New("agentEngine must be codex, claude_code, or pi")
		}
		return engine, nil
	}

	provider := strings.ToLower(strings.TrimSpace(legacyProvider))
	if provider != "" && provider != agentEngineCodex {
		return "", errors.New("legacy provider must be codex")
	}
	profile := strings.TrimPrefix(strings.ToLower(strings.TrimSpace(legacyAgentProfile)), "@")
	switch profile {
	case "", "triage", "codex", "bugfix", "design":
		return agentEngineCodex, nil
	default:
		return "", errors.New("legacy agentProfile is no longer supported")
	}
}

func agentEngineFromHistoricalPayload(agentEngine, legacyProvider, legacyAgentProfile string) (string, error) {
	if strings.TrimSpace(agentEngine) != "" {
		return normalizeAgentEngineInput(agentEngine, "", "")
	}
	provider := strings.ToLower(strings.TrimSpace(legacyProvider))
	if provider == "" || provider == agentEngineCodex {
		// Old custom profiles were always executed by Codex. Keep their sessions
		// readable without preserving the deleted prompt-profile behavior.
		return agentEngineCodex, nil
	}
	return "", errors.New("historical agent payload has an unsupported provider")
}

func isSupportedAgentEngine(agentEngine string) bool {
	switch agentEngine {
	case agentEngineCodex, agentEngineClaudeCode, agentEnginePi:
		return true
	default:
		return false
	}
}

func agentEngineCapability(agentEngine string) (string, error) {
	switch agentEngine {
	case agentEngineCodex:
		return "codex", nil
	case agentEngineClaudeCode:
		return "claudeCode", nil
	case agentEnginePi:
		return "pi", nil
	default:
		return "", errors.New("unsupported agentEngine")
	}
}

func requireCodexWorkflowAgentEngine(agentEngine, legacyProvider, legacyAgentProfile string) (string, error) {
	engine, err := normalizeAgentEngineInput(agentEngine, legacyProvider, legacyAgentProfile)
	if err != nil {
		return "", err
	}
	if engine != agentEngineCodex {
		return "", errors.New("this system workflow currently requires agentEngine codex")
	}
	return engine, nil
}

func normalizeAgentSessionRuntimeTask(requiredCapabilities, payload json.RawMessage) (json.RawMessage, json.RawMessage, error) {
	var payloadObject map[string]json.RawMessage
	if err := json.Unmarshal(payload, &payloadObject); err != nil {
		return nil, nil, fmt.Errorf("agent_session payload must be a JSON object: %w", err)
	}
	agentEngine, err := optionalJSONStringField(payloadObject, "agentEngine")
	if err != nil {
		return nil, nil, err
	}
	legacyProvider, err := optionalJSONStringField(payloadObject, "provider")
	if err != nil {
		return nil, nil, err
	}
	legacyProfile, err := optionalJSONStringField(payloadObject, "agentProfile")
	if err != nil {
		return nil, nil, err
	}
	engine, err := normalizeAgentEngineInput(agentEngine, legacyProvider, legacyProfile)
	if err != nil {
		return nil, nil, err
	}
	engineJSON, _ := json.Marshal(engine)
	payloadObject["agentEngine"] = engineJSON
	delete(payloadObject, "provider")
	delete(payloadObject, "agentProfile")
	payload, err = json.Marshal(payloadObject)
	if err != nil {
		return nil, nil, fmt.Errorf("marshal agent_session payload: %w", err)
	}

	var capabilities map[string]json.RawMessage
	if err := json.Unmarshal(requiredCapabilities, &capabilities); err != nil {
		return nil, nil, fmt.Errorf("agent_session requiredCapabilities must be a JSON object: %w", err)
	}
	capability, err := agentEngineCapability(engine)
	if err != nil {
		return nil, nil, err
	}
	for key, raw := range capabilities {
		if !isAgentEngineCapabilityKey(key) {
			continue
		}
		var enabled bool
		if err := json.Unmarshal(raw, &enabled); err != nil {
			return nil, nil, fmt.Errorf("agent engine capability %s must be boolean", key)
		}
		if enabled && key != capability {
			return nil, nil, fmt.Errorf("requiredCapabilities conflicts with agentEngine %s", engine)
		}
	}
	capabilities[capability] = json.RawMessage("true")
	requiredCapabilities, err = json.Marshal(capabilities)
	if err != nil {
		return nil, nil, fmt.Errorf("marshal agent_session requiredCapabilities: %w", err)
	}
	return requiredCapabilities, payload, nil
}

func optionalJSONStringField(object map[string]json.RawMessage, key string) (string, error) {
	raw, exists := object[key]
	if !exists || len(raw) == 0 || string(raw) == "null" {
		return "", nil
	}
	var value string
	if err := json.Unmarshal(raw, &value); err != nil {
		return "", fmt.Errorf("agent_session payload %s must be a string", key)
	}
	return value, nil
}

File: packages/core/src/agent-engines.ts
import type { AgentEngine, AgentEngineCapability, AgentEngineCatalogItem, AgentSession } from "./types";

export const agentEngineCapabilities: Record<AgentEngine, AgentEngineCapability> = {
  codex: "codex",
  claude_code: "claudeCode",
  pi: "pi",
};

const legacyCodexProfiles = new Set(["triage", "bugfix", "design"]);

function engineToken(value: unknown) {
  return typeof value === "string"
    ? value.trim().toLowerCase().replace(/[^a-z0-9]+/g, "_").replace(/^_+|_+$/g, "")
    : "";
}

export function parseAgentEngine(value: unknown): AgentEngine | undefined {
  const token = engineToken(value);
  if (token === "codex" || legacyCodexProfiles.has(token)) return "codex";
  if (token === "claude" || token === "claude_code" || token === "claudecode") return "claude_code";
  if (token === "pi" || token === "pie") return "pi";
  return undefined;
}

export function agentEngineForSession(session: Pick<AgentSession, "agentEngine" | "provider" | "agentProfile">): AgentEngine {
  return parseAgentEngine(session.agentEngine)
    || parseAgentEngine(session.provider)
    || parseAgentEngine(session.agentProfile)
    || "codex";
}

export function agentCapabilityForEngine(value: unknown): AgentEngineCapability | undefined {
  const engine = parseAgentEngine(value);
  return engine ? agentEngineCapabilities[engine] : undefined;
}

export function isFixedAgentEngineCatalogItem(agent: AgentEngineCatalogItem): boolean {
  const engine = parseAgentEngine(agent.id);
  return Boolean(
    engine
    && engine === agent.id
    && agent.capability === agentEngineCapabilities[engine]
    && agent.mention === agentEngineMention(engine),
  );
}

export function agentRequiredCapabilities(agent: Pick<AgentEngineCatalogItem, "id">): Partial<Record<AgentEngineCapability, boolean>> {
  return { [agentEngineCapabilities[agent.id]]: true };
}

export function agentEngineMention(value: unknown): "@codex" | "@claude" | "@pi" | "" {
  const engine = parseAgentEngine(value);
  if (!engine) return "";
  if (engine === "claude_code") return "@claude";
  return engine === "pi" ? "@pi" : "@codex";
}

export function agentEngineDisplayName(value: unknown): "Codex" | "Claude Code" | "Pi" {
  const engine = parseAgentEngine(value) || "codex";
  if (engine === "claude_code") return "Claude Code";
  return engine === "pi" ? "Pi" : "Codex";
}

export function engineSessionRef(session: Pick<AgentSession, "engineSessionRef" | "codexThreadId">) {
  return session.engineSessionRef || session.codexThreadId || "";
}

export function engineRunRef(session: Pick<AgentSession, "engineRunRef" | "codexTurnId">) {
  return session.engineRunRef || session.codexTurnId || "";
}
